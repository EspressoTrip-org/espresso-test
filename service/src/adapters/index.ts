export * as policy_cache from './policy-cache';
