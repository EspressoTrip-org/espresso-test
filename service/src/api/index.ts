export * as organization from './organizations';
export * as compilation from './compilation';
export * as policies from './policies';
export * as tokens from './tokens';
export * as roles from './roles';
export * as users from './users';
