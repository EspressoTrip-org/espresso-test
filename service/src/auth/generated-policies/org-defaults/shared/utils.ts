import * as cardinal from '@journeyapps-platform/types-cardinal';
import { createOwnerPolicy } from '../owner';
import { createDeveloperPolicy } from '../developer';

export enum MANAGED_POLICY {
  OWNER = 'owner',
  DEVELOPER = 'developer',
  LOCKED = 'locked'
}

export type ManagedPolicyResponse = Record<MANAGED_POLICY.OWNER | MANAGED_POLICY.DEVELOPER, cardinal.Policy>;

export const createDefaultPoliciesForOrg = (org_id: string): ManagedPolicyResponse => {
  return {
    [MANAGED_POLICY.OWNER]: createOwnerPolicy(org_id),
    [MANAGED_POLICY.DEVELOPER]: createDeveloperPolicy(org_id)
  };
};

export type CreateDefaultRolesPolicyMap = Partial<Record<MANAGED_POLICY, string>>;

export const createDefaultRolesForOrg = (org_id: string, policies: CreateDefaultRolesPolicyMap) => {
  const owner_policy = policies[MANAGED_POLICY.OWNER];
  const owner_role: cardinal.Role = {
    name: cardinal.MANAGED_ROLE.OWNER,
    org_id: org_id,
    policy_ids: [owner_policy!]
  };

  const developer_policy = policies[MANAGED_POLICY.DEVELOPER];
  const developer_role: cardinal.Role = {
    name: cardinal.MANAGED_ROLE.DEVELOPER,
    org_id: org_id,
    policy_ids: [developer_policy!]
  };

  return [owner_role, developer_role];
};
