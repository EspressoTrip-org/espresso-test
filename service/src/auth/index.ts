export * from './generated-policies';
export * from './authorizers';
export * from './context';
export * from './viewer';
