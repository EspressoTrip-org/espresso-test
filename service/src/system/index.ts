import { InterServiceClient } from '@journeyapps-platform/sdk-inter-service';
import * as entity_resolver from '@journeyapps-platform/recon-entity-resolver';
import * as invalidator from '@journeyapps-platform/recon-event-invalidator';
import * as micro from '@journeyapps-platform/micro';
import * as adapters from '../adapters';
import * as kafka_utils from './kafka';
import * as resolver from '../recon';
import { MongoDB } from './mongo';
import * as kafka from 'kafkajs';
import ENV from '../env';

export type CreateSystemParams = {
  mocks: string[];
  accounts: {
    endpoint: string;
    client_id: string;
    client_secret: string;
  };
  mongo: micro.mongo.MongoDBAdapterOptions;
  kafka: micro.kafka.KafkaConnectionOptions;
  node_env: string;
};

export class System extends micro.system.MicroSystem {
  context: CreateSystemParams;

  mongo: MongoDB;
  kafka: micro.kafka.KafkaClientProvisioner;
  producer: kafka.Producer;
  resolver: entity_resolver.EntityResolver;

  policy_cache: adapters.policy_cache.PolicyCache;

  constructor(params: CreateSystemParams) {
    super();

    this.context = params;

    const inter_service = new InterServiceClient(params.accounts);

    this.mongo = this.withLifecycle(new MongoDB(params.mongo), {
      start(component) {
        return component.connect();
      },
      stop(component) {
        return component.close();
      }
    });

    this.kafka = this.withLifecycle(micro.kafka.createClientProvisioner(params.kafka), {
      start(component) {
        return kafka_utils.ensureTopics({
          kafka: component(),
          production: params.node_env === 'production'
        });
      }
    });

    this.producer = this.withLifecycle(
      this.kafka('producer').producer({
        createPartitioner: kafka.Partitioners.LegacyPartitioner
      }),
      {
        start(component) {
          return component.connect();
        },
        stop(component) {
          return component.disconnect();
        }
      }
    );

    this.resolver = this.withLifecycle(
      resolver.createResolver(
        this.mongo,
        params.mocks.includes('recon')
          ? undefined
          : {
              external: {
                endpoint: params.accounts.endpoint,
                client: inter_service
              }
            }
      ),
      {}
    );

    this.withLifecycle(
      invalidator.createEventInvalidator({
        resolver: this.resolver,
        kafka: this.kafka('invalidator'),
        modules: [invalidator.resource_operation_events_module, invalidator.cardinal_module]
      }),
      {
        start: async (invalidator) => {
          await invalidator.start();
          await this.resolver.init();
        },
        stop(invalidator) {
          return invalidator.stop();
        }
      }
    );

    this.policy_cache = adapters.policy_cache.createInMemoryPolicyCache();
  }

  static fromENV(env: typeof ENV) {
    return new System({
      mocks: env.MOCK,
      accounts: {
        endpoint: env.JWT_ISSUER,
        client_id: env.ACCOUNTS_CLIENT_ID,
        client_secret: env.ACCOUNTS_CLIENT_SECRET
      },
      mongo: {
        uri: env.MONGO_URI,
        database: env.MONGO_DATABASE,
        username: env.MONGO_USERNAME,
        password: env.MONGO_PASSWORD
      },
      kafka: {
        brokers: env.KAFKA_BROKERS,
        username: env.KAFKA_USERNAME,
        password: env.KAFKA_PASSWORD,
        mechanism: env.KAFKA_MECHANISM as any
      },
      node_env: env.NODE_ENV
    });
  }
}
