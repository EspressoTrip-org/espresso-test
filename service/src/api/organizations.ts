import * as cardinal from '@journeyapps-platform/types-cardinal';
import * as micro from '@journeyapps-platform/micro';
import * as producer from './producer';
import * as policies from './policies';
import * as tokens from './tokens';
import { System } from '../system';
import * as roles from './roles';
import * as users from './users';
import * as auth from '../auth';
import * as bson from 'bson';

type PolicyProvisionParams = {
  org_id: string;
};

export const provisionStandardPoliciesAndRolesForOrg = async (system: System, params: PolicyProvisionParams) => {
  const actor: cardinal.SystemActor = {
    type: cardinal.ActorType.System
  };

  const generated_policies = auth.createDefaultPoliciesForOrg(params.org_id);
  const new_policies = await Promise.all(
    Object.values(generated_policies).map((policy) => {
      return policies.createPolicy(system, {
        actor: actor,
        policy: {
          ...policy,
          managed: true
        }
      });
    })
  );

  const policy_map = new_policies.reduce((acc: auth.CreateDefaultRolesPolicyMap, policy) => {
    acc[policy.name as auth.MANAGED_POLICY] = policy.id;
    return acc;
  }, {});

  const generated_roles = auth.createDefaultRolesForOrg(params.org_id, policy_map);
  return await Promise.all(
    generated_roles.map((role) => {
      return roles.createRole(system, {
        actor,
        role: {
          ...role,
          managed: true
        }
      });
    })
  );
};

type EmitInvalidationEventParams = {
  org: micro.db.SerializedId & cardinal.Organization;
};

/**
 * Emit user assignment changed events which act as cache invalidation events.
 */
export const emitInvalidationEvents = async (system: System, params: EmitInvalidationEventParams) => {
  const org_users = await users.listUsersForOrganization(system.mongo, {
    org_id: params.org.id
  });

  await producer.produceAuthEvents(system, {
    actor: {
      type: cardinal.ActorType.System
    },
    events: org_users.map((user) => {
      return {
        type: cardinal.UserAuthEventType.USER_ASSIGNMENTS_CHANGED,
        payload: user
      };
    })
  });
};

type UpsertOrganizationParams = {
  org: micro.db.SerializedId & cardinal.Organization;
  created_by?: string;
};

export const upsertOrganization = async (system: System, params: UpsertOrganizationParams) => {
  const existing = await system.mongo.organizations.findOne(micro.mongo.getById(params.org.id));

  await system.mongo.organizations.updateOne(
    {
      _id: new bson.ObjectId(params.org.id)
    },
    {
      $set: {
        name: params.org.name,
        locked: params.org.locked
      }
    },
    { upsert: true }
  );

  if (existing) {
    const locked_changed = params.org.locked !== existing.locked;
    if (locked_changed) {
      await emitInvalidationEvents(system, params);
    }

    return;
  }

  const roles = await provisionStandardPoliciesAndRolesForOrg(system, {
    org_id: params.org.id
  });

  // If there is a created_by field on the event then assign the Owner role to the user who
  // created this organization
  if (!params.created_by) {
    return;
  }

  try {
    const user = await users.getUser(system.mongo, params.created_by);
    const owner_role = roles.find((role) => role.name === cardinal.MANAGED_ROLE.OWNER);
    if (!owner_role) {
      return;
    }

    await users.updateUserAssignment(system, {
      actor: {
        type: cardinal.ActorType.System
      },
      assignment: 'role_ids',
      assignment_id: owner_role.id,
      id: user.id,
      op: 'add'
    });
  } catch (e) {
    micro.logger.error(`user ${params.created_by} (creator of org) could not be found`);
  }
};

export const deleteOrganization = async (system: System, params: { id: string }) => {
  await system.mongo.organizations.deleteOne({
    _id: new bson.ObjectId(params.id)
  });

  const all_policies = await policies.listScopedPoliciesForOrganization(system.mongo, {
    org_id: params.id
  });
  const all_roles = await roles.listRolesByOrganization(system.mongo, {
    org_id: params.id
  });
  const all_tokens = await tokens.listScopedTokensForOrganization(system.mongo, {
    org_id: params.id
  });

  // This is a pretty inefficient way to clean up but org deletion will be very in-frequent
  // and as such this is a nice way to re-use already tested logic to do the cleanup

  for (const token of all_tokens) {
    await tokens.deleteToken(system, {
      actor: {
        type: cardinal.ActorType.System
      },
      id: token.id
    });
  }

  for (const role of all_roles) {
    await roles.deleteRole(system, {
      actor: {
        type: cardinal.ActorType.System
      },
      id: role.id
    });
  }

  for (const policy of all_policies) {
    await policies.deletePolicy(system, {
      actor: {
        type: cardinal.ActorType.System
      },
      id: policy.id
    });
  }
};
