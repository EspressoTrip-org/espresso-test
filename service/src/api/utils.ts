import { EntityFilter } from '@journeyapps-platform/recon-entity-resolver';
import * as cardinal from '@journeyapps-platform/types-cardinal';
import * as micro from '@journeyapps-platform/micro';
import { System } from '../system';
import * as mongo from 'mongodb';
import * as bson from 'bson';
import { Actor } from '@journeyapps-platform/types-cardinal';

export type WithActor<T> = T & {
  actor: Actor;
};

export type OptionalPermissions<T> = T & {
  policies?: cardinal.RawPolicy[];
};

export type RequirePermissions<T> = T & {
  policies: cardinal.RawPolicy[];
};

export const getByIdOrThrow = async <T extends micro.db.Id>(collection: mongo.Collection<T>, id: string) => {
  const res = await collection.findOne(micro.mongo.getById(id) as mongo.Filter<T>);
  if (!res) {
    throw new micro.errors.ResourceNotFound(collection.collectionName, id);
  }
  return micro.mongo.toJson(res);
};

type WrapQueryParams<T> = OptionalPermissions<cardinal.V2ListResourceParams> & {
  query: mongo.Filter<T>;
  model: string;
};
export const wrapQueryWithPermissions = async <T extends micro.db.Id & { org_id?: string; user_id?: string }>(
  system: System,
  params: WrapQueryParams<T>
): Promise<any> => {
  if (!params.policies) {
    return params.query;
  }
  const entity_filter: EntityFilter = {};
  if (params.id && !micro.mongo.isQueryFilter(params.id)) {
    entity_filter.id = params.id;
  }
  if (params.org_id && !micro.mongo.isQueryFilter(params.org_id)) {
    entity_filter.organization_id = params.org_id;
  }
  if (params.user_id && !micro.mongo.isQueryFilter(params.user_id)) {
    entity_filter.user_id = params.user_id;
  }

  const ids = await system.resolver.resolveResourceIds({
    policies: params.policies,
    model: params.model,
    entity_filter
  });

  return {
    $and: [
      {
        _id: {
          $in: ids.map((id) => new bson.ObjectId(id))
        }
      },
      params.query
    ]
  };
};
