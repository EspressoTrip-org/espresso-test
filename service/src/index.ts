import '@journeyapps-platform/micro/register';

import * as micro from '@journeyapps-platform/micro';
import * as routes from './router/routes/index';
import { Router } from './router/router';
import * as consumers from './consumers/index';
import { System } from './system/index';
import * as auth from './auth/index';
import fastify from 'fastify';
import env from './env';

const server = fastify();

const system = System.fromENV(env);
const jwt_decoder = auth.createJWTDecoder(system, {
  issuer_url: env.JWT_ISSUER,
  verify: !env.DEV_MODE_DO_NOT_ENABLE_IN_PRODUCTION_OR_YOU_WILL_BE_FIRED
});

const token_decoder = auth.createTokenDecoder(system);
const dev_decoder = auth.createDevDecoder();

const decoders = [jwt_decoder, token_decoder];
if (env.DEV_MODE_DO_NOT_ENABLE_IN_PRODUCTION_OR_YOU_WILL_BE_FIRED) {
  decoders.push(dev_decoder);
}

server.register(
  Router.plugin({
    routes: [...routes.v1_routes, ...routes.v2_routes, ...micro.router.createProbeRoutes()],
    contextProvider: async (payload) => {
      return {
        system: system,
        viewer: await auth.createViewer({
          token: payload.request.headers['authorization'] || '',
          system,
          decoders
        }),
        sha: env.SHA
      };
    }
  })
);

(async () => {
  await system.start();

  await micro.fastify.startServer(server, env.PORT);
  micro.logger.info(`HTTP server running on port ${env.PORT}`);
  await micro.signals.getSystemProbe().ready();

  const [resource_events_consumer, cache_invalidator] = await Promise.all([
    consumers.createResourceOperationEventsConsumer(system),
    consumers.createCacheInvalidator(system)
  ]);

  micro.signals.getTerminationHandler()?.handleTerminationSignal(async () => {
    await Promise.all([resource_events_consumer.disconnect(), cache_invalidator.disconnect()]);
  });
})().catch((err) => {
  micro.logger.error('Something went wrong while booting', err);
  process.exit(1);
});
