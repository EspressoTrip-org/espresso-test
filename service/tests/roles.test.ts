import { beforeAll, afterEach, afterAll, describe, test, expect } from 'vitest';

import * as cardinal from '@journeyapps-platform/types-cardinal';
import { System } from '../src/system';
import * as auth from '../src/auth';
import * as api from '../src/api';
import * as mocks from './mocks';
import * as utils from './utils';
import * as bson from 'bson';
import { AppModelAction, CardinalModel, RoleModelAction } from '@journeyapps-platform/cardinal-catalog';

let system: System;

beforeAll(async () => {
  system = await mocks.createMockedSystem();
  await system.start();
});

afterEach(() => {
  (system.producer as any).clear();
});

afterAll(async () => {
  await system.stop();
});

describe('roles', () => {
  describe('role-crud', () => {
    test('should list roles by scope', async () => {
      await api.roles.createRole(system, {
        actor: cardinal.SYSTEM_ACTOR,
        role: {
          name: 'developer',
          org_id: '123',
          policy_ids: []
        }
      });

      const roles = await api.roles.listRolesByOrganization(system.mongo, {
        org_id: '123'
      });
      expect(utils.removeVolatile(roles)).toMatchSnapshot();
    });

    test('should successfully update a role', async () => {
      const role = await api.roles.createRole(system, {
        actor: cardinal.SYSTEM_ACTOR,
        role: {
          name: 'developer',
          org_id: '123',
          policy_ids: []
        }
      });

      await api.roles.updateRole(system, {
        actor: cardinal.SYSTEM_ACTOR,
        role: {
          ...role,
          name: 'Updated Role'
        }
      });

      const updated_role = await api.roles.getRole(system.mongo, role.id);
      expect(utils.removeVolatile(updated_role)).toMatchSnapshot();
      expect(utils.getEvents(system)).toMatchSnapshot();
    });

    test('deleting a role should update user assignments', async () => {
      const role = await api.roles.createRole(system, {
        actor: cardinal.SYSTEM_ACTOR,
        role: {
          name: 'developer',
          org_id: '123',
          policy_ids: []
        }
      });

      const now = new Date();
      const remaining_role_id = '5e6f5519ede3631176831f4c';
      await system.mongo.roles.insertOne({
        _id: new bson.ObjectId(remaining_role_id),
        name: 'dangling',
        org_id: '123',
        policy_ids: [],
        created_at: now,
        updated_at: now
      });

      const user = await system.mongo.users.insertOne({
        _id: new bson.ObjectId(),

        org_id: '123',
        email: 'test@journeyapps.com',
        role_ids: [role.id, remaining_role_id]
      });

      await api.roles.deleteRole(system, {
        actor: cardinal.SYSTEM_ACTOR,
        id: role.id
      });

      const updated_user = await system.mongo.users.findOne({
        _id: user.insertedId
      });

      expect(updated_user?.role_ids?.length).toBe(1);
      expect(updated_user?.role_ids?.[0]).toBe(remaining_role_id);

      expect(utils.getEvents(system)).toMatchSnapshot();
    });
  });

  describe('role-auth', () => {
    test('should be unauthorized to create role [operation]', async () => {
      const role: cardinal.Role = {
        policy_ids: [],
        name: 'role',
        org_id: '123'
      };

      const context = mocks.createMockedContext(system, []);

      const res = await auth.roles.authorizedToCreateRole({ params: role, context });
      expect(res.authorized).toBeFalsy();
    });

    const common_policy: cardinal.Policy = {
      org_id: '123',
      statements: [
        {
          actions: [AppModelAction.READ],
          resources: [
            {
              scope: '123',
              selector: {
                model: CardinalModel.APP,
                id: '*'
              }
            }
          ]
        }
      ]
    };

    const create_permission: cardinal.Policy = {
      statements: [
        {
          actions: [RoleModelAction.CREATE],
          resources: [
            {
              scope: '123',
              selector: {
                model: CardinalModel.ROLE,
                id: '*'
              }
            }
          ]
        }
      ]
    };

    const read_permission: cardinal.Policy = {
      statements: [
        {
          actions: [AppModelAction.READ],
          resources: [
            {
              scope: '*',
              selector: {
                model: CardinalModel.APP,
                id: '*'
              }
            }
          ]
        }
      ]
    };

    test('should be authorized to create role [operation]', async () => {
      const role: cardinal.Role = {
        policy_ids: [],
        name: 'role',
        org_id: '123'
      };

      const context = mocks.createMockedContext(system, [create_permission]);

      const res = await auth.roles.authorizedToCreateRole({ params: role, context });
      expect(res.authorized).toBeTruthy();
    });

    test('should be unauthorized to create role [scope]', async () => {
      const policy = await api.policies.createPolicy(system, {
        actor: cardinal.SYSTEM_ACTOR,
        policy: {
          ...common_policy,
          org_id: undefined
        }
      });
      const role: cardinal.Role = {
        policy_ids: [policy.id],
        name: 'role',
        org_id: '123'
      };

      const context = mocks.createMockedContext(system, [create_permission, read_permission]);

      const res = await auth.roles.authorizedToCreateRole({ params: role, context });
      expect(res.authorized).toBeFalsy();
    });

    test('should be unauthorized to create role [policy/role scope mismatch]', async () => {
      const policy = await api.policies.createPolicy(system, {
        actor: cardinal.SYSTEM_ACTOR,
        policy: {
          ...common_policy,
          org_id: '1234'
        }
      });
      const role: cardinal.Role = {
        policy_ids: [policy.id],
        name: 'role',
        org_id: '123'
      };

      const context = mocks.createMockedContext(system, [create_permission, read_permission]);

      const res = await auth.roles.authorizedToCreateRole({ params: role, context });
      expect(res.authorized).toBeFalsy();
    });

    test('should be unauthorized to create role [policies]', async () => {
      const policy = await api.policies.createPolicy(system, {
        actor: cardinal.SYSTEM_ACTOR,
        policy: common_policy
      });
      const role: cardinal.Role = {
        policy_ids: [policy.id],
        name: 'role',
        org_id: '123'
      };

      const context = mocks.createMockedContext(system, [create_permission]);

      const res = await auth.roles.authorizedToCreateRole({ params: role, context });
      expect(res.authorized).toBeFalsy();
    });

    test('should be authorized to create role [policies]', async () => {
      const policy = await api.policies.createPolicy(system, {
        actor: cardinal.SYSTEM_ACTOR,
        policy: common_policy
      });
      const role: cardinal.Role = {
        policy_ids: [policy.id],
        name: 'role',
        org_id: '123'
      };

      const context = mocks.createMockedContext(system, [create_permission, read_permission]);

      const res = await auth.roles.authorizedToCreateRole({ params: role, context });
      expect(res.authorized).toBeTruthy();
    });
  });
});
