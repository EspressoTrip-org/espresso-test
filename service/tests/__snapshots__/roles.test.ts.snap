// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`roles > role-crud > deleting a role should update user assignments 1`] = `
[
  {
    "metadata": {
      "actor": {
        "type": "system",
      },
    },
    "payload": {
      "name": "developer",
      "org_id": "123",
      "policy_ids": [],
    },
    "type": "AUTH.ROLE_CREATED",
  },
  {
    "metadata": {
      "actor": {
        "type": "system",
      },
    },
    "payload": {
      "email": "test@journeyapps.com",
      "org_id": "123",
      "role_ids": [
        "5e6f5519ede3631176831f4c",
      ],
    },
    "type": "AUTH.USER_ASSIGNMENTS_CHANGED",
  },
  {
    "metadata": {
      "actor": {
        "type": "system",
      },
    },
    "payload": {
      "name": "developer",
      "org_id": "123",
      "policy_ids": [],
    },
    "type": "AUTH.ROLE_DELETED",
  },
]
`;

exports[`roles > role-crud > should list roles by scope 1`] = `
[
  {
    "name": "developer",
    "org_id": "123",
    "policy_ids": [],
  },
]
`;

exports[`roles > role-crud > should successfully update a role 1`] = `
{
  "name": "Updated Role",
  "org_id": "123",
  "policy_ids": [],
}
`;

exports[`roles > role-crud > should successfully update a role 2`] = `
[
  {
    "metadata": {
      "actor": {
        "type": "system",
      },
    },
    "payload": {
      "name": "developer",
      "org_id": "123",
      "policy_ids": [],
    },
    "type": "AUTH.ROLE_CREATED",
  },
  {
    "metadata": {
      "actor": {
        "type": "system",
      },
    },
    "payload": {
      "name": "Updated Role",
      "org_id": "123",
      "policy_ids": [],
    },
    "type": "AUTH.ROLE_UPDATED",
  },
]
`;
